core/engine/src/vm/call_frame/mod.rs

```diff
@@ -81,22 +80,49 @@ impl CallFrame {
 impl CallFrame {
     /// This is the size of the function prologue.
     ///
-    /// The position of the elements are relative to the [`CallFrame::fp`].
+    /// The position of the elements are relative to the [`CallFrame::fp`] (frame pointer).
+    ///
+    /// ```text
```

Comment:
General thought on this: Would it be better to store this `fp` as a `sp` (although I'm not sure it's a true stack pointer, maybe `lp` for local pointer), and keep the `fp` and make up the memory by computing the argument count via `self.sp - self.fp - Self::FUNCTION_PROLOGUE` or would computing the args count be more inefficient?

Reply:
> Would it be better to store this `fp` as a `sp`

I don't think calling it sp would be correct, sp is implict in the stack vec itself that being the length (pointing to the position that the next value will be put).

Native calling conventions the arguments passed are not part of the call frame of the callee. But it doesn't really matter, since we can make our own convention. :smile: 

The biggest reason for `fp` change is to make reduce the computation needed to access locals before we would have to do `fp + FUNCTION_PROLOGUE + arg_count + local_offset` with this change we only do `fp + local_offset`.

I kind of hinted at the next change, switching to a register based VM :smile: Registers will be used extensively so we want fast access. I think it's a good time to switch before we are too dependent on the stack version, while working on #3037 Modeling data flow of stack based VMs is much harder, and was conflicted whether to continue development, since it would make going switching much harder and it will have to be redone.

Reply:
Yepp! I noticed the register based shift!

Just for clarity, I meant keep `fp` and then `local pointer`/`sp` (whatever we call it) in exchange for `argument_count`. It would then exchanging `restore_fp` in exchange for calculating `argument_count`. I think we still compute the argument_start anyways, which can then just be `fp + FUNCTION_PROLOGUE` over `fp - argument_count`.

```rust
// Current dumbed down version of `CallFrame`
struct DummyFrame {
    fp: u32,
    argument_count: u32
}

// New version
struct NewFrame {
    fp: u32,
    lp: u32
}

// argument_count
impl NewFrame {
    // Trade off would be that we compute arg_count over compute fp
    fn arg_count(&self) -> u32 {
        self.lp - self.arg_start()
    }
    
    // Throwing this in as a utility
    fn arg_start(&self) -> u32 {
        self.fp + FUNCTION_PROLOGUE
    }
}
```

The idea was that maybe we save a bit of compute by not having to use `restore_fp` in exchange for calculating the arg_count. At least that was the thought I had (could be off on that), but thought I'd run it by you nonetheless. Of course, if that would somehow undercut the register shift you had in mind, then completely disregard this.

Reply:
Sorry for the delay. I decided to leave this for after we implement the register VM, added a todo so I don't forget :smiley: .

I renamed the fp into rp (register pointer), calling it register pointer, instead of local pointer for consistency with the registers that will be added into the future.